File: templatemethod/exercise/VideoPreset.java
package templatemethod.exercise;

public abstract class VideoPreset {

    public void applyPreset() {
        enhanceVideoQuality();
        applyColorCorrection();
        enhanceAudioQuality();
        applyFilters();
        renderVideo();
    }

    private void applyFilters() {
    }

    abstract protected void applyColorCorrection();

    private void enhanceVideoQuality() {

    }

    private void enhanceAudioQuality() {

    }

    abstract protected void renderVideo();
}

File: typescript/template/exercise/video-preset.ts
export abstract class VideoPreset {
  applyPreset() {
    this.enhanceVideoQuality();
    this.applyColorCorrection();
    this.enhanceAudioQuality();
    this.applyFilters();
    this.renderVideo();
  }

  private applyFilters() {}

  protected abstract applyColorCorrection(): void;

  private enhanceVideoQuality() {}

  private enhanceAudioQuality() {}

  protected abstract renderVideo(): void;
}
